packages/storage/src/file.js

```diff
@@ -491,63 +505,77 @@ File.prototype.createReadStream = function(options) {
       var start = is.number(options.start) ? options.start : '0';
       var end = is.number(options.end) ? options.end : '';
 
-      reqOpts.headers = {
-        Range: 'bytes=' + (tailRequest ? end : start + '-' + end)
-      };
+      reqOpts.headers.Range =
+        'bytes=' + (tailRequest ? end : start + '-' + end);
     }
 
-    var requestStream = self.requestStream(reqOpts);
-    var validateStream;
+    var requestStream = self.requestStream(reqOpts)
+      .on('error', function(err) {
+        throughStream.destroy(err);
+      })
+      .on('response', function(res) {
+        throughStream.emit('response', res);
+        common.util.handleResp(null, res, null, onResponse);
+      })
+      .on('complete', function(res) {
+        common.util.handleResp(null, res, null, onComplete);
+      })
+      .resume();
+
+    throughStream.on('error', function() {
+      // An error can occur before the request stream has been created
+      // (during authentication).
+      if (requestStream.abort) {
+        requestStream.abort();
+      }
+
+      requestStream.destroy();
+    });
 
     // We listen to the response event from the request stream so that we can...
     //
     //   1) Intercept any data from going to the user if an error occurred.
     //   2) Calculate the hashes from the http.IncomingMessage response stream,
     //      which will return the bytes from the source without decompressing
-    //      gzip'd content. The request stream will do the decompression so the
-    //      user receives the expected content.
-    function onResponse(err, body, res) {
+    //      gzip'd content. We then send it through decompressed, if applicable,
+    //      to the user.
+    function onResponse(err, body, rawResponseStream) {
       if (err) {
-        requestStream.unpipe(throughStream);
-
         // Get error message from the body.
-        res.pipe(concat(function(body) {
+        rawResponseStream.pipe(concat(function(body) {
           err.message = body.toString();
           throughStream.destroy(err);
         }));
 
         return;
       }
 
-      if (!rangeRequest) {
-        validateStream = hashStreamValidation({
-          crc32c: crc32c,
-          md5: md5
-        });
+      var headers = rawResponseStream.toJSON().headers;
+      var isCompressed = headers['content-encoding'] === 'gzip';
 
-        res.pipe(validateStream).on('data', common.util.noop);
-      }
+      rawResponseStream
+        .pipe(shouldRunValidation ? validateStream : through())
+        .pipe(isCompressed ? zlib.createGunzip() : through())
```

Comment:
I think we could optimize this if we were to conditionally pipe here. My understanding is that `through` uses a moderate amount of memory, so ideally we wouldn't use it as a noop. I think `pumpify` might be useful here

Reply:
Good call, will do!